### ISSUE TITLE
retry behaviour

### ISSUE THREAD

**Issue by bamarni** [containous/traefik#1104]
_(follow-up from #1016)_

At the moment the condition for retrying a request is the following : 

``` go
func isNetworkError(status int) bool {
	return status == http.StatusBadGateway || status == http.StatusGatewayTimeout
}
```

Those response status can come either from upstream or be set by the [default oxy error handler](https://github.com/containous/oxy/blob/fcc76b52eb8568540a020b7a99e854d9d752b364/utils/handler.go#L18-L31). To compare this according to [proxy_next_upstream](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_next_upstream) semantics, that makes it like :

    error timeout http_502 http_504 non_idempotent

Personally I find it too open for a default, if upstream sends http 504 for a POST request it still might have altered a resource but didn't manage to serve the response in a timely manner. For instance with PHP-FPM and nginx I believe this is a common cause for 504 responses, and retrying those POST requests could cause duplicate data or other unexpected side effects.

Should we make this configurable? Or maybe make the pattern more strict? (to compare with nginx again, their default is only `error timeout`)

cc @emilevauge @SantoDE

**Comment by jonaz** [comment]
Auch! i always assuemed traefik only retries GET. PUT and POST is really dangerous to retry...

**Comment by timoreimann** [comment]
HAProxy has a similar default: Only [refused connections and timeouts](https://cbonte.github.io/haproxy-dconv/1.8/configuration.html#retries) are eligible for retries.

I agree we should adapt a more safe default and consider making this configurable if there's a need for it.

**Comment by timoreimann** [comment]
Looking at the definition of 502s and 504s more closely, the error codes are meant to be returned by servers _acting as a gateway or proxy_. In that sense, non-intermediaries should not emit such error codes AFAICS.

That's not to say we shouldn't change the behavior.

**Comment by timoreimann** [comment]
I'm planning to work on this one next.

Which path should we follow? I see three options:

1. Change the existing logic to only retry on connection errors and connection timeouts (not counting established connections).
1. Change the existing logic as in 1. but allow to customize the codes on which Traefik should retry (additionally or exclusively).
1. Keep the existing logic and add an option to (additionally or exclusively) retry on connection errors/timeouts.

Personally, I'm in favor of 1. I'd rather not do 2. until/unless we get some feedback that users actually want a means to customize the codes. Otherwise, we risk maintaining code nobody wants.

I'm against 3. because it was never a safe default in the first place; as such, I consider it "broken design" that we should fix.

Thoughts anyone?

@containous/traefik, what do you say?

**Comment by emilevauge** [comment]
@timoreimann I'm also in favor of 1, KISS :)

**Comment by aschen77** [comment]
@timoreimann I have a use case for a web app that is in the process of shutting down to programmatically have Traefik retry the call to another instance. We currently use 502/504 response codes to accomplish this so would be in favor of option 2.

**Comment by timoreimann** [comment]
@aschen77 to make sure I understand you correctly: do you that the application is actively producing 502/504 responses and tells Traefik to retry when it's about to shut down?

**Comment by aschen77** [comment]
@timoreimann Yes that is correct. It is a workaround to support graceful shutdown of the application. We use Marathon for application lifecycle and waiting for a better integration with Traefik to remove the app instance from Traefik upon app shutdown.

**Comment by timoreimann** [comment]
I think a better option for that case would be to implement [graceful termination](https://mesosphere.github.io/marathon/docs/health-checks.html#taskkillgraceperiodseconds) in Marathon. This requires making your application keep listening for requests on reception of a SIGTERM for a short period while Traefik stops forwarding requests when the `TASK_KILLING` state is reached. Traefik does not respect the state yet, but it's easy to modify the vanilla template accordingly: [This](https://github.com/containous/traefik/blob/69c31276f2f22de2336e6935a8861313d8b3f27e/templates/marathon.tmpl#L2-L6) needs an additional `if` block:

```
{{$apps := .Applications}}
[backends]{{range .Tasks}}
    {{ if not ( eq .State "TASK_KILLING" ) }}
        [backends."backend{{getBackend . $apps}}".servers."server-{{.ID | replace "." "-"}}"]
        url = "{{getProtocol . $apps}}://{{getBackendServer . $apps}}:{{getPort . $apps}}"
        weight = {{getWeight . $apps}}+  {{ end }}
    {{end}}
 {{end}}
```

It'd also be easy to upstream this into Traefik.

**Comment by aschen77** [comment]
Thanks for this @timoreimann , this is the proper approach. We will test this and submit a PR to include it in the default Marathon template.

**Comment by timoreimann** [comment]
@aschen77 looking forward to it ☺️

**Comment by ldez** [comment]
Closed by #1549
